In [1]:
import pandas as pd
import numpy as np
import numpy as np

### 0_Preluare date in dataframe-uri pandas

In [2]:
df_energ_log = pd.read_csv('input/EnergLoc.csv')
df_ronuts = pd.read_csv('input/RO_NUTS.csv')

### 0_Vizualizare seturi de date

In [3]:
df_energ_log

Judet  c3511  c3512  c3513  c3514   c3521   c3522   c3523   c3530
0     dj  7.700  0.000  4.138  1.795   0.000   0.000   0.000   4.770
1     ms  0.206  0.000  0.000  0.000   0.000  11.061  11.842   1.422
2     vl  0.071  0.000  0.000  0.000   0.000   0.000   2.312  15.843
3     hr  0.071  0.000  1.428  0.347   0.000   0.000   2.844   7.952
4     nt  0.327  0.000  0.807  0.000   0.000   0.528   0.000   2.400
5     tm  0.582  0.000  2.010  0.195   0.000   0.206   0.380   4.336
6     tl  1.720  0.000  1.193  0.000   0.000   2.316   0.000   3.521
7     tr  1.061  0.000  0.000  0.000   0.000   0.000   0.000   9.198
8     bz  0.495  0.000  0.671  0.246   1.501   0.000   0.000   4.568
9     bv  0.561  2.427  2.400  0.691   0.000   0.451   0.000   1.727
10    vs  0.076  0.000  0.000  0.588   0.000   2.117   0.000   1.871
11    br  0.328  0.000  0.932  0.000   6.919   0.000   0.000   4.872
12    vn  0.302  0.000  0.000  0.000   0.000   0.000   0.000   5.934
13    bt  1.604  0.000  0.000  0.000  11.965   0.000   0.000   7.214
14    bn  0.985  0.000  0.000  0.000   0.000   0.000   0.000   1.352
15    if  0.445  0.000  0.000  0.000   0.739   0.487   1.446   0.000
16    sv  0.671  0.000  0.000  0.000   0.000   0.000   0.894   9.531
17    mh  0.109  0.000  0.000  1.364   0.000   1.185   0.000   3.085
18    bh  0.321  0.000  0.000  0.298   0.000   0.698   0.000   5.938
19    bc  0.786  0.000  0.000  0.000   0.000   0.000   0.000   5.255
20    ot  0.112  0.000  0.000  0.000   0.000   0.000   3.150   0.000
21    mm  0.334  0.000  0.000  0.248   0.000   0.000   0.000   0.000
22    db  0.098  0.000  0.000  0.000   0.000   0.000   0.000   2.753
23    is  0.263  0.000  4.019  0.635   0.000   1.862   0.000   1.679
24    sm  0.885  0.000  0.000  0.000   0.000   0.000   0.000   0.000
25     b  1.277  2.434  1.081  2.601   2.330   1.427   1.901   2.664
26    ar  2.362  0.000  0.000  0.296   0.905   0.678   0.000   0.767
27    hd  2.188  0.000  0.000  0.236   0.000   0.000   0.000   0.000
28    il  0.361  0.000  0.000  0.000   0.000   0.000   0.000   0.000
29    sj  1.236  0.000  0.000  1.248   0.000   1.276   0.000   5.747
30    ct  0.799  0.000  1.708  0.862   0.000   1.323   0.000   4.905
31    cs  1.522  0.000  1.115  0.000   0.000   0.000   0.000   2.820
32    gl  2.000  0.000  0.000  0.745   0.000   0.479   0.000   0.000
33    gj  0.138  0.000  0.000  0.000   0.000   0.000   0.000   2.757
34    sb  0.375  4.360  0.544  0.239   0.000   0.000   0.000   1.869
35    cl  0.083  0.000  0.000  0.000   0.000   0.882   0.000   2.368
36    cj  0.448  0.000  2.160  0.639   0.000   0.638   0.000   1.825
37    ab  0.100  0.000  0.000  0.892   1.488   0.000   0.000   0.000
38    ph  0.757  2.619  2.853  0.147   0.663   0.749   0.000   2.881
39    ag  1.045  0.000  0.437  0.000   0.000   0.000   0.000   4.885
40    gr  2.337  0.000  0.000  0.000  10.528   0.000   0.000   0.000

In [4]:
df_ronuts

IndicativJudet             NumeJudet          Regiune MacroRegiune
0              ab                  Alba           Centru          RO1
1              ag                 Arges              Sud          RO3
2              ar                  Arad             Vest          RO4
3               b  Municipiul Bucuresti  Bucuresti-Ilfov          RO3
4              bc                 Bacau         Nord-Est          RO2
5              bh                 Bihor        Nord-Vest          RO1
6              bn       Bistrita-Nasaud        Nord-Vest          RO1
7              br                Braila          Sud-Est          RO2
8              bt              Botosani         Nord-Est          RO2
9              bv                Brasov           Centru          RO1
10             bz                 Buzau          Sud-Est          RO2
11             cj                  Cluj        Nord-Vest          RO1
12             cl              Calarasi              Sud          RO3
13             cs         Caras-Severin             Vest          RO4
14             ct             Constanta          Sud-Est          RO2
15             cv               Covasna           Centru          RO1
16             db             Dambovita              Sud          RO3
17             dj                  Dolj         Sud-Vest          RO4
18             gj                  Gorj         Sud-Vest          RO4
19             gl                Galati          Sud-Est          RO2
20             gr               Giurgiu              Sud          RO3
21             hd             Hunedoara             Vest          RO4
22             hr              Harghita           Centru          RO1
23             if                 Ilfov  Bucuresti-Ilfov          RO3
24             il              Ialomita              Sud          RO3
25             is                  Iasi         Nord-Est          RO2
26             mh             Mehedinti         Sud-Vest          RO4
27             mm             Maramures        Nord-Vest          RO1
28             ms                 Mures           Centru          RO1
29             nt                 Neamt         Nord-Est          RO2
30             ot                   Olt         Sud-Vest          RO4
31             ph               Prahova              Sud          RO3
32             sb                 Sibiu           Centru          RO1
33             sj                 Salaj        Nord-Vest          RO1
34             sm             Satu Mare        Nord-Vest          RO1
35             sv               Suceava         Nord-Est          RO2
36             tl                Tulcea          Sud-Est          RO2
37             tm                 Timis             Vest          RO4
38             tr             Teleorman              Sud          RO3
39             vl                Valcea         Sud-Vest          RO4
40             vn               Vrancea          Sud-Est          RO2
41             vs                Vaslui         Nord-Est          RO2

### 1_Sa se calculeze coeficientii de localizare la nivel de judet ca medie a coeficientilor subactivitatilor
- Se creaza un dataframe nou in care copiem pe prima coloana judetul si pe a doua coloana media pe linie a subactivitatilor
    - luam valorile cu subactivitatile folosind iloc --> toate elementele de la coloana 2 pana la sfarsit
    - calculam media pe linie (axis = 1) pentru fiecare entry in parte
- Rezultatul se salveaza intr-un Cerinta1.csv

In [5]:
df_coef_loc = pd.DataFrame()

df_coef_loc['Judet'] = df_energ_log['Judet']
df_coef_loc['Coef_loc'] = df_energ_log.iloc[:, 1:].mean(axis=1)

df_coef_loc.to_csv('output/Cerinta1.csv')

In [6]:
df_coef_loc

Judet  Coef_loc
0     dj  2.300375
1     ms  3.066375
2     vl  2.278250
3     hr  1.580250
4     nt  0.507750
5     tm  0.963625
6     tl  1.093750
7     tr  1.282375
8     bz  0.935125
9     bv  1.032125
10    vs  0.581500
11    br  1.631375
12    vn  0.779500
13    bt  2.597875
14    bn  0.292125
15    if  0.389625
16    sv  1.387000
17    mh  0.717875
18    bh  0.906875
19    bc  0.755125
20    ot  0.407750
21    mm  0.072750
22    db  0.356375
23    is  1.057250
24    sm  0.110625
25     b  1.964375
26    ar  0.626000
27    hd  0.303000
28    il  0.045125
29    sj  1.188375
30    ct  1.199625
31    cs  0.682125
32    gl  0.403000
33    gj  0.361875
34    sb  0.923375
35    cl  0.416625
36    cj  0.713750
37    ab  0.310000
38    ph  1.333625
39    ag  0.795875
40    gr  1.608125

### 2_Sa se extraga judetele care pentru coeficientul C3511 este mai mare ca 1 si sa se salveze intr-un fisier
- Se poate face o masca peste setul de date in care se filtreaza judetele cu C3511 > 1
- Rezultatul se salveaza in Cerinta2.csv

In [7]:
mask = df_energ_log['c3511'] > 1
df_energ_log[mask].to_csv('output/Cerinta2.csv')

In [8]:
df_energ_log[mask]

Judet  c3511  c3512  c3513  c3514   c3521  c3522  c3523  c3530
0     dj  7.700  0.000  4.138  1.795   0.000  0.000  0.000  4.770
6     tl  1.720  0.000  1.193  0.000   0.000  2.316  0.000  3.521
7     tr  1.061  0.000  0.000  0.000   0.000  0.000  0.000  9.198
13    bt  1.604  0.000  0.000  0.000  11.965  0.000  0.000  7.214
25     b  1.277  2.434  1.081  2.601   2.330  1.427  1.901  2.664
26    ar  2.362  0.000  0.000  0.296   0.905  0.678  0.000  0.767
27    hd  2.188  0.000  0.000  0.236   0.000  0.000  0.000  0.000
29    sj  1.236  0.000  0.000  1.248   0.000  1.276  0.000  5.747
31    cs  1.522  0.000  1.115  0.000   0.000  0.000  0.000  2.820
32    gl  2.000  0.000  0.000  0.745   0.000  0.479  0.000  0.000
39    ag  1.045  0.000  0.437  0.000   0.000  0.000  0.000  4.885
40    gr  2.337  0.000  0.000  0.000  10.528  0.000  0.000  0.000

### 3_Sa se salveze judetul si subactivitatea cea mai mare din acel judet 
- Se creaza un dataframe nou in care se pun judetele 
- Se foloseste functia idxmax pe linie pentru a da fetch la numele coloanei pe care se afla valoare maxima
- Rezultatul se salveaza in Cerinta3.csv

In [9]:
df_subactiv_max = pd.DataFrame()

df_subactiv_max['Judet'] = df_energ_log['Judet']
df_subactiv_max['Subactiv_max'] = df_energ_log.iloc[:, 1:].idxmax(axis=1)

df_subactiv_max.to_csv('output/Cerinta3.csv')

In [10]:
df_subactiv_max

Judet Subactiv_max
0     dj        c3511
1     ms        c3523
2     vl        c3530
3     hr        c3530
4     nt        c3530
5     tm        c3530
6     tl        c3530
7     tr        c3530
8     bz        c3530
9     bv        c3512
10    vs        c3522
11    br        c3521
12    vn        c3530
13    bt        c3521
14    bn        c3530
15    if        c3523
16    sv        c3530
17    mh        c3530
18    bh        c3530
19    bc        c3530
20    ot        c3523
21    mm        c3511
22    db        c3530
23    is        c3513
24    sm        c3511
25     b        c3530
26    ar        c3511
27    hd        c3511
28    il        c3511
29    sj        c3530
30    ct        c3530
31    cs        c3530
32    gl        c3511
33    gj        c3530
34    sb        c3512
35    cl        c3530
36    cj        c3513
37    ab        c3521
38    ph        c3530
39    ag        c3530
40    gr        c3521

### 4_Sa se calculeze coeficientii medii de localizare pe nivel de regiuni 
- Facem un merge intre df_energ_loc si df_ronuts pentru a lua regiunile
- Dupa ce am facut merge facem un groupyby pe regiune unde calculam mediile astfel:
    - Facem groupby-ul folosind mean() pentru a face media valorilor numerice
    - Pentru ca groupBy-ul cu mean sa functioneze trebuie sa selectam doar coloanele numerice pentru ca nu se poate face groupby pe string-uri 
- Salvam dataframe-ul in fisierul Cerinta4.csv

In [11]:
df_coef_reg = df_energ_log
df_coef_reg = df_energ_log.merge(df_ronuts[['IndicativJudet', 'Regiune']], left_on='Judet', right_on='IndicativJudet')
df_coef_reg = df_coef_reg.groupby('Regiune')[df_coef_reg.select_dtypes('number').columns].mean()
df_coef_reg.to_csv('output/Cerinta4.csv')

In [12]:
df_coef_reg

c3511     c3512     c3513     c3514     c3521     c3522  \
Regiune                                                                       
Bucuresti-Ilfov  0.861000  1.217000  0.540500  1.300500  1.534500  0.957000   
Centru           0.262600  1.357400  0.874400  0.433800  0.297600  2.302400   
Nord-Est         0.621167  0.000000  0.804333  0.203833  1.994167  0.751167   
Nord-Vest        0.701500  0.000000  0.360000  0.405500  0.000000  0.435333   
Sud              0.820286  0.374143  0.470000  0.021000  1.598714  0.233000   
Sud-Est          0.940667  0.000000  0.750667  0.308833  1.403333  0.686333   
Sud-Vest         1.626000  0.000000  0.827600  0.631800  0.000000  0.237000   
Vest             1.663500  0.000000  0.781250  0.181750  0.226250  0.221000   

                  c3523     c3530  
Regiune                            
Bucuresti-Ilfov  1.6735  1.332000  
Centru           2.9372  2.594000  
Nord-Est         0.1490  4.658333  
Nord-Vest        0.0000  2.477000  
Sud              0.0000  3.155000  
Sud-Est          0.0000  3.966667  
Sud-Vest         1.0924  5.291000  
Vest             0.0950  1.980750

### 5_Sa se salveze salveze per regiune judetele cu cei mai mari coeficienti in subactivitati
- Fac un join intre df_energ_loc si df_ronuts pentru a introduce regiunile in setul de date
    - Index-ul joinului va fi reprezentat de field-ul 'Judet'
- Fac un groupby custom pe regiuni in care fiecare element rezultat va fi index-ul (judetul) unde valoarea subactivitatii este maxima  

#### Functia care se apeleaza in groupby-ul custom
- Dataframe-ul df are ca si index pe fiecare linie judetul 
- Functia se apeleaza de atatea ori cate regiuni avem (de atatea ori cate valori avem in field-ul din groupby)
    - La fiecare apel, df va fi un slice din setul de date care reprezinta o grupare pe regiuni 
- Fiecare valoare pe care o returnam reprezinta valoarea maxima dintre toate judetele per subactivitate
    - ex:
        -        c3511  c3512  c3513  c3514  c3521  c3522  c3523  c3530
        -  Judet                                                        
        -  if     0.445  0.000  0.000  0.000  0.739  0.487  1.446  0.000
        -  b      1.277  2.434  1.081  2.601  2.330  1.427  1.901  2.664
    - vom aveam b,b,b,b,b,b,b,b pentru ca b > if peste tot

In [35]:
def custom_apply_function(df:pd.DataFrame):
    print(df)
    max = df.idxmax()
    return max

!! Setez ca index 'Judet' ca sa pot in custom_apply_function() sa extrag judetul !!

In [36]:
df_aggreg = df_energ_log.merge(df_ronuts[['IndicativJudet', 'Regiune']], left_on='Judet', right_on='IndicativJudet', how='inner').set_index('Judet')
#drop useless indicativJudet column from resulted merge
df_aggreg.drop('IndicativJudet', axis=1, inplace=True)

In [37]:
df_aggreg = df_aggreg.groupby('Regiune').apply(custom_apply_function, include_groups=False)

       c3511  c3512  c3513  c3514  c3521  c3522  c3523  c3530
Judet                                                        
if     0.445  0.000  0.000  0.000  0.739  0.487  1.446  0.000
b      1.277  2.434  1.081  2.601  2.330  1.427  1.901  2.664
       c3511  c3512  c3513  c3514  c3521   c3522   c3523  c3530
Judet                                                          
ms     0.206  0.000  0.000  0.000  0.000  11.061  11.842  1.422
hr     0.071  0.000  1.428  0.347  0.000   0.000   2.844  7.952
bv     0.561  2.427  2.400  0.691  0.000   0.451   0.000  1.727
sb     0.375  4.360  0.544  0.239  0.000   0.000   0.000  1.869
ab     0.100  0.000  0.000  0.892  1.488   0.000   0.000  0.000
       c3511  c3512  c3513  c3514   c3521  c3522  c3523  c3530
Judet                                                         
nt     0.327    0.0  0.807  0.000   0.000  0.528  0.000  2.400
vs     0.076    0.0  0.000  0.588   0.000  2.117  0.000  1.871
bt     1.604    0.0  0.000  0.000  11.965  0.000  0.

In [40]:
df_aggreg.to_csv('output/Cerinta5.csv')

In [41]:
df_aggreg

c3511 c3512 c3513 c3514 c3521 c3522 c3523 c3530
Regiune                                                        
Bucuresti-Ilfov     b     b     b     b     b     b     b     b
Centru             bv    sb    bv    ab    ab    ms    ms    hr
Nord-Est           bt    nt    is    is    bt    vs    sv    sv
Nord-Vest          sj    bn    cj    sj    bn    sj    bn    bh
Sud                gr    ph    ph    ph    gr    cl    tr    tr
Sud-Est            gl    tl    ct    ct    br    tl    tl    vn
Sud-Vest           dj    dj    dj    dj    dj    mh    ot    vl
Vest               ar    tm    tm    ar    ar    ar    tm    tm